src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -254,4 +267,88 @@ void justClose() {
             SCOPED_MEMORY_ACCESS.closeScope(this);
         }
     }
+
+    /**
+     * A functional interface modelling the cleanup action associated with a scope.
+     */
+    interface CleanupAction {
+        void cleanup();
+        CleanupAction dup();
+
+        /** Dummy cleanup action */
+        CleanupAction DUMMY = new CleanupAction() {
+            @Override
+            public void cleanup() {
+                // do nothing
+            }
+
+            @Override
+            public CleanupAction dup() {
+                return this;
+            }
+        };
+    }
+
+    /**
+     * A stateful cleanup action; this action can only be called at most once. The implementation
+     * guarantees this invariant even when multiple threads race to call the {@link #cleanup()} method.
+     */
+    static abstract class BasicCleanupAction implements CleanupAction {
+
+        static final VarHandle CALLED;
+
+        static {
+            try {
+                CALLED = MethodHandles.lookup().findVarHandle(BasicCleanupAction.class, "called", boolean.class);
+            } catch (Throwable ex) {
+                throw new ExceptionInInitializerError(ex);
+            }
+        }
+
+        private boolean called = false;
+
+        abstract void run();
+
+        public final void cleanup() {
+            disable();
+            run();
+        };
+
+        @Override
+        public CleanupAction dup() {
+            disable();
+            return new BasicCleanupAction() {
+                @Override
+                void run() {
+                    BasicCleanupAction.this.run();
+                }
+            };
+        }
+
+        final void disable() {
+            if (!CALLED.compareAndSet(this, false, true)) {
+                // This can never happen under normal circumstances. The only case where this can happen is when
+                // when two cleaners race to cleanup the same scope. It is never possible to have a race
+                // between explicit/implicit close because all the scope terminal operations have
+                // reachability fences which prevent a scope to be deemed unreachable before we are done
+                // marking the original cleanup action as "dead".
+                throw new IllegalStateException("Already cleaned");
```

Comment:
I recommend not throwing, it may cause noise for the developer when they observe through say flags or events that exceptions are being thrown.